File: commonutils/src/main/java/net/maxsmr/commonutils/android/ServiceUtils.java
package net.maxsmr.commonutils.android;

import android.app.ActivityManager;
import android.app.AlarmManager;
import android.app.Notification;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.SystemClock;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import net.maxsmr.commonutils.android.notification.NotificationActionInfo;
import net.maxsmr.commonutils.android.notification.NotificationController;
import net.maxsmr.commonutils.android.notification.NotificationInfo;

public final class ServiceUtils {

    private final static Logger logger = LoggerFactory.getLogger(ServiceUtils.class);

    private ServiceUtils() {
        throw new AssertionError("no instances.");
    }

    public static <S extends Service> boolean isServiceRunning(@NonNull Context context, @NonNull Class<S> serviceClass) {
        ActivityManager manager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName())) {
                return true;
            }
        }
        return false;
    }

    public static <S extends Service> boolean isServiceForeground(@NonNull Context context, @NonNull Class<S> serviceClass) {
        ActivityManager manager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        for (ActivityManager.RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
            if (serviceClass.getName().equals(service.service.getClassName()) && service.foreground) {
                return true;
            }
        }
        return false;
    }

    private static <S extends Service> void startNoCheck(@NonNull Context context, @NonNull Class<S> serviceClass) {
        Intent service = new Intent(context, serviceClass);
        context.startService(service);
    }

    public static <S extends Service> void start(@NonNull Context context, @NonNull Class<S> serviceClass) {
        logger.debug("start(), serviceClass= " + serviceClass);
        if (!isServiceRunning(context, serviceClass)) {
            restartNoCheck(context, serviceClass, null);
        }
    }

    private static <S extends Service> void restartNoCheck(@NonNull Context context, @NonNull Class<S> serviceClass, @Nullable S serviceInst) {
        logger.debug("restart(), serviceClass=" + serviceClass);
        stopNoCheck(context, serviceClass, serviceInst);
        startNoCheck(context, serviceClass);
    }

    public static <S extends Service> void restart(@NonNull Context context, @NonNull Class<S> serviceClass, @Nullable S serviceInst) {
        logger.debug("restart(), serviceClass=" + serviceClass);
        stop(context, serviceClass, serviceInst);
        startNoCheck(context, serviceClass);
    }

    private static <S extends Service> void stopNoCheck(@NonNull Context context, @NonNull Class<S> serviceClass, @Nullable S serviceInst) {
//        if (serviceInst != null) {
//            serviceInst.onDestroy();
//        }
        Intent service = new Intent(context, serviceClass);
        context.stopService(service);
    }

    public static <S extends Service> void stop(@NonNull Context context, @NonNull Class<S> serviceClass, @Nullable S serviceInst) {
        logger.debug("stop(), serviceClass=" + serviceClass + ", serviceInst=" + serviceInst);
        if (isServiceRunning(context, serviceClass)) {
            stopNoCheck(context, serviceClass, serviceInst);
        }
    }

    public static <S extends Service> void cancelDelay(@NonNull Context context, @NonNull Class<S> serviceClass) {
        logger.debug("cancelDelay(), serviceClass=" + serviceClass);
        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(context, serviceClass);
        PendingIntent pendingIntent = PendingIntent.getService(context, 0, intent, 0);
        alarmManager.cancel(pendingIntent);
    }

    public static <S extends Service> void restartDelay(@NonNull Context context, @NonNull Class<S> serviceClass, long delay) {
        logger.debug("restartDelay(), serviceClass=" + serviceClass + ", delay=" + delay);

        if (delay < 0) {
            throw new IllegalArgumentException("incorrect delay: " + delay);
        }
        stop(context, serviceClass, null);
        startDelayNoCheck(context, serviceClass, delay);
    }

    private static <S extends Service> void restartDelayNoCheck(@NonNull Context context, @NonNull Class<S> serviceClass, long delay) {

        if (delay < 0) {
            throw new IllegalArgumentException("incorrect delay: " + delay);
        }

        stopNoCheck(context, serviceClass, null);
        startDelayNoCheck(context, serviceClass, delay);
    }

    public static <S extends Service> void startDelayNoCheck(@NonNull Context context, @NonNull Class<S> serviceClass, long delay) {

        cancelDelay(context, serviceClass);

        AlarmManager alarmManager = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        Intent intent = new Intent(context, serviceClass);
        PendingIntent pendingIntent = PendingIntent.getService(context, 0, intent, 0);
        alarmManager.set(AlarmManager.ELAPSED_REALTIME_WAKEUP, SystemClock.elapsedRealtime() + delay, pendingIntent);
    }

    public static <S extends Service> void startDelay(@NonNull Context context, @NonNull Class<S> serviceClass, long delay) {
        logger.debug("startDelay(), serviceClass=" + serviceClass + ", delay=" + delay);
        if (!isServiceRunning(context, serviceClass)) {
            restartDelayNoCheck(context, serviceClass, delay);
        }
    }

    public static void startServiceForeground(@NonNull Service service, int id, @NonNull NotificationInfo notificationInfo) {
        logger.debug("startServiceForeground(), service=" + service + ", id=" + id + ", notificationInfo=" + notificationInfo);
        startServiceForeground(service, id, notificationInfo.contentIntent, notificationInfo.tickerText, notificationInfo.contentTitle, notificationInfo.text, notificationInfo.iconResId, notificationInfo.actionInfos);
    }

    @Nullable
    public static Notification startServiceForeground(@NonNull Service service, int id, @Nullable Intent contentIntent, @Nullable String ticker, @Nullable String title, @Nullable String text, @DrawableRes int iconRes, NotificationActionInfo... actionInfos) {
        logger.debug("startServiceForeground(), service=" + service + ", id=" + id + ", contentIntent=" + contentIntent + ", ticker=" + ticker + ", title=" + title + ", text=" + text + ", iconRes=" + iconRes);
        if (!isServiceForeground(service, service.getClass())) {

            NotificationController.getInstance().removeNotification(id);

            NotificationInfo info = new NotificationInfo();
            info.id = id;
            info.autoCancel = false;
            info.onlyUpdate = false;
            info.ongoing = true;
            info.contentIntent = contentIntent;
            info.actionInfos = actionInfos;
            info.tickerText = ticker;
            info.contentTitle = title;
            info.text = text;

            Notification noti = NotificationController.getInstance().createAndAddNotification(info);
            service.startForeground(id, noti);
            NotificationController.getInstance().updateNotification(id, noti);
            return noti;
        }
        return NotificationController.getInstance().getNotification(id);
    }

    public static <S extends Service> void stopServiceForeground(@NonNull S service, int id) {
        logger.debug("stopServiceForeground(), service=" + service + ", id=" + id);
        if (isServiceForeground(service, service.getClass())) {
            service.stopForeground(true);
            NotificationController.getInstance().removeNotification(id);
        }
    }

    public static <C extends ServiceConnection, S extends Service> boolean bindService(@NonNull Context ctx, @NonNull Class<S> serviceClass, @NonNull C serviceConnection, int flags) {
        logger.debug("bindService(), serviceClass=" + serviceClass + ", serviceConnection=" + serviceConnection + ", flags=" + flags);
        if (!ctx.bindService(new Intent(ctx, serviceClass), serviceConnection, flags)) {
            logger.error("binding to service " + serviceClass + " failed");
            stop(ctx, serviceClass, null);
            start(ctx, serviceClass);
            return false;
        }
        return true;
    }

    public static <C extends ServiceConnection> void unbindService(@NonNull Context context, @NonNull C serviceConnection) {
        logger.debug("unbindService(), serviceConnection=" + serviceConnection);
        context.unbindService(serviceConnection);
    }


}

File: commonutils/src/main/kotlin/net/maxsmr/commonutils/service/ServiceUtils.kt
package net.maxsmr.commonutils.service

import android.annotation.SuppressLint
import android.app.ActivityManager
import android.app.PendingIntent
import android.app.Service
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.text.TextUtils
import net.maxsmr.commonutils.cancelAlarm
import net.maxsmr.commonutils.isAtLeastNougat
import net.maxsmr.commonutils.isAtLeastOreo
import net.maxsmr.commonutils.isAtLeastS
import net.maxsmr.commonutils.logger.BaseLogger
import net.maxsmr.commonutils.logger.holder.BaseLoggerHolder
import net.maxsmr.commonutils.logger.holder.BaseLoggerHolder.Companion.logException
import net.maxsmr.commonutils.setAlarm

private val logger = BaseLoggerHolder.instance.getLogger<BaseLogger>("ServiceUtils")

@JvmOverloads
fun <S : Service> isServiceRunning(
    context: Context,
    serviceClass: Class<S>,
    isForeground: Boolean? = null
): Boolean {
    val manager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager?
        ?: throw RuntimeException("ActivityManager is null")
    return manager.getRunningServices(Integer.MAX_VALUE).find { service ->
        serviceClass.name == service.service.className &&
                (isForeground == null || service.foreground == isForeground)
    } != null
}

@JvmOverloads
fun <S : Service> start(
    context: Context,
    serviceClass: Class<S>,
    args: Bundle? = null,
    action: String? = null,
    startForeground: Boolean = false
): StartResult {
    if (!isServiceRunning(context, serviceClass)) {
        return if (startNoCheck(context, serviceClass, args, action, startForeground)) {
            StartResult.STARTED
        } else {
            StartResult.NOT_STARTED_FAILED
        }
    }
    return StartResult.NOT_STARTED_ALREADY_RUNNING
}

@JvmOverloads
fun <S : Service> restart(
    context: Context,
    serviceClass: Class<S>,
    args: Bundle? = null,
    action: String? = null,
    startForeground: Boolean = false
): Boolean {
    stop(context, serviceClass)
    return startNoCheck(context, serviceClass, args, action, startForeground)
}

fun <S : Service> stop(context: Context, serviceClass: Class<S>): Boolean {
    logger.d("stop, serviceClass=$serviceClass")
    if (isServiceRunning(context, serviceClass)) {
        return stopNoCheck(context, serviceClass)
    }
    return true
}

fun <S : Service> restartDelay(
    context: Context,
    serviceClass: Class<S>,
    delay: Long,
    requestCode: Int = 0,
    isForeground: Boolean = false,
    packageName: String = context.packageName,
    args: Bundle? = null,
    action: String? = null,
    flags: Int = PendingIntent.FLAG_CANCEL_CURRENT,
    shouldWakeUp: Boolean = true
): Boolean {
    logger.d("restartDelay, serviceClass=$serviceClass, delay=$delay")
    require(delay >= 0) { "Incorrect delay: $delay" }
    stop(context, serviceClass)
    return startDelayNoCheck(
        context,
        serviceClass,
        delay,
        requestCode,
        flags,
        isForeground,
        packageName,
        args,
        action,
        shouldWakeUp
    )
}


fun <S : Service> startDelay(
    context: Context,
    serviceClass: Class<S>,
    delay: Long,
    requestCode: Int = 0,
    isForeground: Boolean = false,
    packageName: String = context.packageName,
    args: Bundle? = null,
    action: String? = null,
    flags: Int = PendingIntent.FLAG_CANCEL_CURRENT,
    shouldWakeUp: Boolean = true
): Boolean {
    logger.d("startDelay, serviceClass=$serviceClass, delay=$delay")
    if (!isServiceRunning(context, serviceClass)) {
        return restartDelayNoCheck(
            context,
            serviceClass,
            delay,
            requestCode,
            flags,
            isForeground,
            packageName,
            args,
            action,
            shouldWakeUp
        )
    }
    return false
}

@JvmOverloads
fun <S : Service> cancelDelay(
    context: Context,
    serviceClass: Class<S>,
    requestCode: Int = 0,
    flags: Int = PendingIntent.FLAG_CANCEL_CURRENT,
    isForeground: Boolean = false,
    packageName: String = context.packageName,
    args: Bundle? = null,
    action: String? = null
) {
    logger.d("cancelDelay, serviceClass=$serviceClass")
    val pendingIntent = createServicePendingIntent(
        context,
        serviceClass,
        requestCode,
        flags,
        isForeground,
        packageName,
        args,
        action
    )
    cancelAlarm(context, pendingIntent)
}

fun <C : ServiceConnection, S : Service> bindService(
    context: Context,
    serviceClass: Class<S>,
    serviceConnection: C,
    args: Bundle? = null,
    action: String? = null,
    flags: Int = 0,
    restartIfFailed: Boolean = false
): Boolean {
    logger.d("bindService, serviceClass=$serviceClass, serviceConnection=$serviceConnection, flags=$flags")
    var bounded = false
    try {
        bounded = context.bindService(
            createServiceIntent(
                context.packageName,
                serviceClass,
                args,
                action
            ), serviceConnection, flags
        )
    } catch (e: Exception) {
        logException(logger, e, "bindService")
    }

    if (!bounded) {
        logger.e("Binding to service $serviceClass failed")
        if (restartIfFailed) {
            restart(context, serviceClass)
        }
        return false
    }
    return true
}

fun <C : ServiceConnection> unbindService(context: Context, serviceConnection: C) {
    logger.d("unbindService, serviceConnection=$serviceConnection")
    try {
        context.unbindService(serviceConnection)
    } catch (e: Exception) {
        logException(logger, e, "unbindService")
    }
}

@JvmOverloads
fun createServiceIntent(
    packageName: String,
    serviceClass: Class<out Service>,
    args: Bundle? = null,
    action: String? = null
): Intent {
    require(!TextUtils.isEmpty(packageName)) { "Empty package name: $packageName" }
    val componentName = ComponentName(packageName, serviceClass.name)
    val serviceIntent = if (!action.isNullOrEmpty()) {
        Intent(action)
    } else {
        Intent()
    }.apply {
        args?.let {
            putExtras(it)
        }
    }
    serviceIntent.component = componentName
    return serviceIntent
}

@JvmOverloads
fun createServicePendingIntent(
    context: Context,
    serviceClass: Class<out Service>,
    requestCode: Int = 0,
    flags: Int,
    isForeground: Boolean = false,
    packageName: String = context.packageName,
    args: Bundle? = null,
    action: String? = null
): PendingIntent {
    val intent = createServiceIntent(
        packageName,
        serviceClass,
        args,
        action
    )
    val modifiedFlags = withMutabilityFlag(flags, false)
    return if (isForeground && isAtLeastOreo()) {
        PendingIntent.getForegroundService(
            context,
            requestCode,
            intent,
            modifiedFlags
        )
    } else {
        PendingIntent.getService(
            context,
            requestCode,
            intent,
            modifiedFlags
        )
    }
}

/**
 * @param startForeground true if intended to start in foreground for 8+
 */
@SuppressLint("NewApi")
@JvmOverloads
fun <S : Service> startNoCheck(
    context: Context,
    serviceClass: Class<S>,
    args: Bundle? = null,
    action: String? = null,
    startForeground: Boolean = false
): Boolean {
    val i = createServiceIntent(context.packageName, serviceClass, args, action)
    return try {
        if (startForeground && isAtLeastOreo()) {
            context.startForegroundService(i) != null
        } else {
            context.startService(i) != null
        }
    } catch (e: Exception) {
        // на >=O выбросится если вызов был с опозданием после определённого интервала при уходе аппа в bg
        logException(logger, e, "Cannot start service with intent $i")
        false
    }
}

fun withMutabilityFlag(flags: Int, mutable: Boolean): Int {
    val mutableFlag = if (mutable) {
        if (isAtLeastS()) PendingIntent.FLAG_MUTABLE else 0
    } else {
        PendingIntent.FLAG_IMMUTABLE
    }
    return flags or mutableFlag
}

fun Service.stopForegroundCompat(removeNotification: Boolean) {
    if (removeNotification) {
        stopForeground(true)
    } else {
        if (isAtLeastNougat()) {
            stopForeground(Service.STOP_FOREGROUND_DETACH)
        } else {
            stopForeground(false)
        }
    }
}

private fun <S : Service> stopNoCheck(context: Context, serviceClass: Class<S>): Boolean {
    val service = Intent(context, serviceClass)
    return context.stopService(service)
}

private fun <S : Service> restartDelayNoCheck(
    context: Context,
    serviceClass: Class<S>,
    delay: Long,
    requestCode: Int = 0,
    flags: Int = PendingIntent.FLAG_CANCEL_CURRENT,
    isForeground: Boolean = false,
    packageName: String = context.packageName,
    args: Bundle? = null,
    action: String? = null,
    shouldWakeUp: Boolean = true
): Boolean {

    require(delay >= 0) { "Incorrect delay: $delay" }

    stopNoCheck(context, serviceClass)
    return startDelayNoCheck(
        context,
        serviceClass,
        delay,
        requestCode,
        flags,
        isForeground,
        packageName,
        args,
        action,
        shouldWakeUp
    )
}

private fun <S : Service> startDelayNoCheck(
    context: Context,
    serviceClass: Class<S>,
    delay: Long,
    requestCode: Int = 0,
    flags: Int = PendingIntent.FLAG_CANCEL_CURRENT,
    isForeground: Boolean = false,
    packageName: String = context.packageName,
    args: Bundle? = null,
    action: String? = null,
    shouldWakeUp: Boolean = true
): Boolean {
    cancelDelay(context, serviceClass, requestCode)
    return setAlarm(
        context,
        createServicePendingIntent(
            context,
            serviceClass,
            requestCode,
            flags,
            isForeground,
            packageName,
            args,
            action
        ),
        delay,
        shouldWakeUp
    )
}

enum class StartResult {
    NOT_STARTED_ALREADY_RUNNING,
    NOT_STARTED_FAILED,
    STARTED
}